### PR TITLE
Don't capture `stdout` when launching subshells in `nix repl`

### PR DESCRIPTION
**Pull request by virusdave**
In [this commit](https://github.com/NixOS/nix/commit/9b9e703df41d75949272059f9b8bc8b763e91fce), the behavior of the `:s` and `:u` commands in `nix repl` changed by capturing `stdout` of the forked process.  This doesn't seem to be an intentional change, since the captured output is discarded.

This change reverts the behavior by having `nix repl`'s invocations of `nix` no longer redirect and capture `stdout` of the forked process.

### CODE REVIEW

**Comment by virusdave** [comment]
I think this fixes https://github.com/NixOS/nix/issues/4884 .

```
$ EDITOR=vim nix repl '<nixpkgs>'
:e hello
# vim opens with the `hello` derivation correctly
```

**Comment by virusdave** [comment]
@thufschmitt since no good deed goes unpunished, can I get your thoughts on this one as well?

File: src/libutil/util.cc
Comment (virusdave) [inline src/libutil/util.cc:1102]: I'm not a big fan of this name, and would rather have the "no capture" function be just plain `runProgram` and name the capturing version something else, but that's a more invasive change, so I'll defer to the maintainers/reviewers here on what they prefer.

File: src/libcmd/repl.cc
Comment (thufschmitt) [inline src/libcmd/repl.cc:120]: Apart from the very untelling name, I think you could directly use `runProgram2` here, right?

**Comment by virusdave** [comment]
By duplicating the "capture exception and grab edit code" logic that
doesn't exist in `runProgram2`? Sure, if that's preferred, let me know.

**Comment by thufschmitt** [comment]
> By duplicating the "capture exception and grab edit code" logic that doesn't exist in `runProgram2`?

I might be missing something, but I'm under the impression that the exit code is only used to re-throw a similar exception a few lines later. Which means that the whole logic is "capture exception and grab edit code to immediately rebuild and throw the same (roughly) exception". Which sounds like a complex way of expressing the identity function :wink:

**Comment by virusdave** [comment]
> > By duplicating the "capture exception and grab edit code" logic that doesn't exist in `runProgram2`?
> 
> I might be missing something, but I'm under the impression that the exit code is only used to re-throw a similar exception a few lines later. Which means that the whole logic is "capture exception and grab edit code to immediately rebuild and throw the same (roughly) exception". Which sounds like a complex way of expressing the identity function 😉

Ha, you are indeed correct.  I completely glossed over this, apparently.  I'll simplify this change to do what you're suggesting.  Thanks!

**Comment by virusdave** [comment]
> > > By duplicating the "capture exception and grab edit code" logic that doesn't exist in `runProgram2`?
> > 
> > 
> > I might be missing something, but I'm under the impression that the exit code is only used to re-throw a similar exception a few lines later. Which means that the whole logic is "capture exception and grab edit code to immediately rebuild and throw the same (roughly) exception". Which sounds like a complex way of expressing the identity function 😉
> 
> Ha, you are indeed correct. I completely glossed over this, apparently. I'll simplify this change to do what you're suggesting. Thanks!

Done!

With latest changes:
```
dave @ davembp [/tmp]
$ nix-shell -p 'with pkgs; nix.overrideAttrs (self: super: {patches = fetchpatch {url="https://github.com/virusdave/nix/commit/32978e9c00b7287a6fef624aec8fa854d8062878.patch"; sha256="sha256-KuFThyYAOES/17lyzJLv55tMdeC43Oq60AVvcKj4g2g=";};})'

dave @ davembp [nix-shell:/tmp]
$ nix repl
Welcome to Nix 2.8.1. Type :? for help.

nix-repl> :l <nixpkgs>
Added 16467 variables.

nix-repl> :u hello     
dave @ davembp [nix-shell:/tmp]

$ echo 'Output still works!'
Output still works!

dave @ davembp [nix-shell:/tmp]
$ 
exit

```

**Review by thufschmitt — APPROVED** [review]
Simple and concise, like that :)

Thanks!

**Pull request merged**
